### PR TITLE
ITS-3424 gracefully handling 404s from PDV.

### PR DESCRIPTION
**Pull request by rambabu-posa**

### CODE REVIEW

File: app/controllers/CheckDetailsController.scala [hunk 1]
```diff
@@ -117,9 +139,33 @@ class CheckDetailsController @Inject()(
...
+      case _ @ PDVSuccessResponse(pdvData: PDVResponseData) =>
+        if (pdvData.validationStatus.equals("success")) {
+          checkDetailsMatchingSuccess(pdvData, idData, mode, sessionWithNINO, origin)
+        } else {
+          logger.warn(s"PDV matched failed: ${pdvData.validationStatus}")
+          auditService.findYourNinoPDVMatchFailed(Some(pdvData), origin)
+          Redirect(routes.InvalidDataNINOHelpController.onPageLoad(mode = mode)).withSession(sessionWithNINO)
+        }
+      case _ @ PDVNotFoundResponse(_) =>
+        Redirect(routes.InvalidDataNINOHelpController.onPageLoad(mode = mode)).withSession(sessionWithNINO)
```
Comment (ajchilts) [inline app/controllers/CheckDetailsController.scala:158]: I think we will need to audit the fact the match was not successful here too.

File: app/services/IndividualDetailsService.scala [hunk 2]
```diff
@@ -64,13 +63,22 @@ class IndividualDetailsServiceImpl @Inject()(
...
+  override def getIdData(pdvResponse: PDVResponse)(implicit hc: HeaderCarrier): Future[Either[IndividualDetailsError, IndividualDetails]] = {
+    pdvResponse match {
+      case PDVSuccessResponse(pdvData : PDVResponseData) =>
+        getIndividualDetails(IndividualDetailsNino(pdvData.personalDetails match {
+          case Some(data) => data.nino.nino
+          case None =>
+            logger.warn("No Personal Details found in PDV data, likely validation failed")
+            EmptyString
+        })).value
+      case _ =>
```
Comment (ajchilts) [inline app/services/IndividualDetailsService.scala:75]: I think we should check specifically for the PDV not found response here?

File: app/services/PersonalDetailsValidationService.scala [hunk 3]
```diff
@@ -128,14 +130,12 @@ class PersonalDetailsValidationService @Inject()(connector: PersonalDetailsValid
...
-      pdvData <- createPDVDataFromPDVMatch(body)
-    } yield pdvData match {
-      case data: PDVResponseData =>
-        data
-      case _ =>
-        throw new Exception("No PDV data found")
+      pdv <- createPDVDataFromPDVMatch(body)
+    } yield pdv match {
+      case pdvResponse: PDVResponse => pdvResponse
+      case _ => throw new Exception("No PDV data found")
```
Comment (ajchilts) [inline app/services/PersonalDetailsValidationService.scala:138]: This error message is not technically correct as a 404 (not  found) from PDV would be contained within pdvResponse. 

A more suitable error would be 'PDV response could not be parsed'.

**Review by ajchilts — CHANGES_REQUESTED** [review]

File: app/controllers/CheckDetailsController.scala [hunk 4]
```diff
@@ -98,12 +99,33 @@ class CheckDetailsController @Inject()(
...
+  private def checkDetailsFailureJourney(pdvResponse: PDVResponse,
+                                         idDataError: IndividualDetailsError,
+                                         mode: Mode,
+                                         sessionWithNINO: Session,
+                                         origin: Option[String])(implicit headerCarrier: HeaderCarrier): Result =
+    pdvResponse match {
+      case _ @ PDVSuccessResponse(pdvData: PDVResponseData) =>
+        handleCheckDetailsFailureJourney(idDataError, mode, sessionWithNINO, origin, pdvData)
+      case pdvData @ PDVNotFoundResponse(_) =>
+        // TODO
```
[the remark this answers is not in the document]
Comment (ajchilts) [reply app/controllers/CheckDetailsController.scala:111]: I have asked Alison with regards to capturing a third state in the audit event.

**Review by ajchilts — CHANGES_REQUESTED** [review]

File: app/services/IndividualDetailsService.scala (hunk 2 above)
Comment (rambabu-posa) [reply app/services/IndividualDetailsService.scala:75]: other than PDV success, all fall into `case _ =>` scenario.

**Review by ajchilts — DISMISSED** [review]

File: test/controllers/CheckDetailsControllerSpec.scala [hunk 5]
```diff
@@ -306,6 +306,30 @@ class CheckDetailsControllerSpec extends SpecBase with SummaryListFluency {
...
+        val app = applicationBuilder(userAnswers = Some(emptyUserAnswers))
+          .overrides(
+            inject.bind[PersonalDetailsValidationService].toInstance(mockPersonalDetailsValidationService),
+            inject.bind[CheckDetailsService].toInstance(mockCheckDetailsService),
+            inject.bind[AuditService].toInstance(auditService)
+          )
+          .build()
+
+        when(mockPersonalDetailsValidationService.createPDVDataFromPDVMatch(any())(any()))
+          .thenReturn(Future.successful(mockPDVResponseDataSuccess))
```
Comment (ajchilts) [inline test/controllers/CheckDetailsControllerSpec.scala:319]: Is this returning a not found? Reads like a success response?

Comment (rambabu-posa) [reply test/controllers/CheckDetailsControllerSpec.scala:319]: oh right, I fixed it. now

**Review by ajchilts — DISMISSED** [review]

**Review by ajchilts — APPROVED** [review]

**Pull request merged**